File: agent/packages/schema/src/dugu_v2.ts
import { type Static, Type } from "@sinclair/typebox";

import { type ValidationResult, validate } from "./validate.js";

export const DuguV2SkillIdV1 = Type.Union([
  Type.Literal("eclipse"),
  Type.Literal("self_cure"),
  Type.Literal("penetrate"),
  Type.Literal("shroud"),
  Type.Literal("reverse"),
]);
export type DuguV2SkillIdV1 = Static<typeof DuguV2SkillIdV1>;

export const DuguTaintTierV1 = Type.Union([
  Type.Literal("immediate"),
  Type.Literal("temporary"),
  Type.Literal("permanent"),
]);
export type DuguTaintTierV1 = Static<typeof DuguTaintTierV1>;

export const DuguV2SkillCastV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    target: Type.Optional(Type.String({ minLength: 1 })),
    skill: DuguV2SkillIdV1,
    tick: Type.Integer({ minimum: 0 }),
    taint_tier: Type.Optional(DuguTaintTierV1),
    hp_loss: Type.Number({ minimum: 0 }),
    qi_loss: Type.Number({ minimum: 0 }),
    qi_max_loss: Type.Number({ minimum: 0 }),
    permanent_decay_rate_per_min: Type.Number({ minimum: 0 }),
    returned_zone_qi: Type.Number({ minimum: 0 }),
    reveal_probability: Type.Number({ minimum: 0, maximum: 1 }),
    animation_id: Type.String({ minLength: 1 }),
    particle_id: Type.String({ minLength: 1 }),
    sound_recipe_id: Type.String({ minLength: 1 }),
    icon_texture: Type.String({ minLength: 1 }),
  },
  { additionalProperties: false },
);
export type DuguV2SkillCastV1 = Static<typeof DuguV2SkillCastV1>;

export const DuguSelfCureProgressV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    hours_used: Type.Number({ minimum: 0 }),
    daily_hours_after: Type.Number({ minimum: 0, maximum: 6 }),
    gain_percent: Type.Number({ minimum: 0 }),
    insidious_color_percent: Type.Number({ minimum: 0, maximum: 100 }),
    morphology_percent: Type.Number({ minimum: 0, maximum: 100 }),
    self_revealed: Type.Boolean(),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type DuguSelfCureProgressV1 = Static<typeof DuguSelfCureProgressV1>;

export const DuguReverseTriggeredV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    affected_targets: Type.Integer({ minimum: 0 }),
    burst_damage: Type.Number({ minimum: 0 }),
    returned_zone_qi: Type.Number({ minimum: 0 }),
    juebi_delay_ticks: Type.Optional(Type.Integer({ minimum: 0 })),
    center: Type.Tuple([Type.Number(), Type.Number(), Type.Number()]),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type DuguReverseTriggeredV1 = Static<typeof DuguReverseTriggeredV1>;

export function validateDuguV2SkillCastV1Contract(data: unknown): ValidationResult {
  return validate(DuguV2SkillCastV1, data);
}

export function validateDuguSelfCureProgressV1Contract(data: unknown): ValidationResult {
  return validate(DuguSelfCureProgressV1, data);
}

export function validateDuguReverseTriggeredV1Contract(data: unknown): ValidationResult {
  return validate(DuguReverseTriggeredV1, data);
}

File: server/src/schema/dugu_v2.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DuguV2SkillIdV1 {
    Eclipse,
    SelfCure,
    Penetrate,
    Shroud,
    Reverse,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DuguTaintTierV1 {
    Immediate,
    Temporary,
    Permanent,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DuguV2SkillCastV1 {
    pub caster: String,
    pub target: Option<String>,
    pub skill: DuguV2SkillIdV1,
    pub tick: u64,
    pub taint_tier: Option<DuguTaintTierV1>,
    pub hp_loss: f32,
    pub qi_loss: f32,
    pub qi_max_loss: f32,
    pub permanent_decay_rate_per_min: f32,
    pub returned_zone_qi: f32,
    pub reveal_probability: f32,
    pub animation_id: String,
    pub particle_id: String,
    pub sound_recipe_id: String,
    pub icon_texture: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DuguSelfCureProgressV1 {
    pub caster: String,
    pub hours_used: f32,
    pub daily_hours_after: f32,
    pub gain_percent: f32,
    pub insidious_color_percent: f32,
    pub morphology_percent: f32,
    pub self_revealed: bool,
    pub tick: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DuguReverseTriggeredV1 {
    pub caster: String,
    pub affected_targets: u32,
    pub burst_damage: f32,
    pub returned_zone_qi: f32,
    pub juebi_delay_ticks: Option<u64>,
    pub center: [f64; 3],
    pub tick: u64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dugu_v2_skill_cast_serializes_snake_case_skill_and_tier() {
        let payload = DuguV2SkillCastV1 {
            caster: "player:a".to_string(),
            target: Some("player:b".to_string()),
            skill: DuguV2SkillIdV1::Eclipse,
            tick: 7,
            taint_tier: Some(DuguTaintTierV1::Permanent),
            hp_loss: 20.0,
            qi_loss: 40.0,
            qi_max_loss: 0.0,
            permanent_decay_rate_per_min: 0.001,
            returned_zone_qi: 39.6,
            reveal_probability: 0.03,
            animation_id: "bong:dugu_needle_throw".to_string(),
            particle_id: "bong:dugu_taint_pulse".to_string(),
            sound_recipe_id: "dugu_needle_hiss".to_string(),
            icon_texture: "bong:textures/gui/skill/dugu_eclipse.png".to_string(),
        };
        let json = serde_json::to_string(&payload).unwrap();
        assert!(json.contains(r#""skill":"eclipse""#));
        assert!(json.contains(r#""taint_tier":"permanent""#));
    }
}
